wordhelper/anagram.py

```diff
@@ -1,24 +1,77 @@
 
 def anagram(pattern):
-	words_by_character_frequency = CURRENT_DICTIONARY.words_by_character_frequency
-	wildcard_stripped = list(filter(lambda c: c != "?", pattern))
-	wildcard_stripped_histogram = get_character_histogram(wildcard_stripped)
-	anagrams = []
-	expanded_histograms = expand_wildcard(wildcard_stripped_histogram, len(pattern) - len(wildcard_stripped))
-	for histogram in expanded_histograms:
-		if histogram in words_by_character_frequency:
-			anagrams += words_by_character_frequency[histogram]
-	return anagrams
-
-def expand_wildcard(character_histogram, num_wildcards):
-	expanded_histograms = []
-	if num_wildcards == 0:
-		expanded_histograms.append(histogram_to_string(character_histogram))
-	else:
+	pattern = AnagramPattern(pattern=pattern)
+	return pattern.find()
+
+def anagram_multi(pattern):
+	pattern = AnagramPattern(pattern=pattern)
+	return pattern.find_multi()
+
+class AnagramPattern:
+	def __init__(self, pattern=None, character_histogram=None, num_wildcards=None):
+		if character_histogram == None:
+			wildcard_stripped = list(filter(lambda c: c != "?", pattern))
+			character_histogram = get_character_histogram(wildcard_stripped)
+			num_wildcards = len(pattern) - len(wildcard_stripped)
+
+		self.character_histogram = character_histogram
+		self.num_wildcards = num_wildcards
+
+	def expand_wildcard(self):
+		return self.expand_wildcard_helper(0, self.num_wildcards)
+
+	def expand_wildcard_helper(self, starting_from, num_wildcards_left):
+		expanded_histograms = []
+		if num_wildcards_left == 0:
+			expanded_histograms.append(histogram_to_string(self.character_histogram))
+		else:
+			for i in range(starting_from, 26):
+				self.character_histogram[i] += 1
+				expanded_histograms += self.expand_wildcard_helper(i, num_wildcards_left - 1)
+				self.character_histogram[i] -= 1
+		return expanded_histograms
+
+	def subtract_word(self, word):
+		# Creates a copy of self.character_histogram
+		new_character_histogram = list(self.character_histogram)
+		new_num_wildcards = self.num_wildcards
+		word_character_histogram = get_character_histogram(word)
 		for i in range(26):
-			character_histogram[i] += 1
-			expanded_histograms += expand_wildcard(character_histogram, num_wildcards - 1)
-			character_histogram[i] -= 1
-	return expanded_histograms
\ No newline at end of file
+			new_character_histogram[i] -= word_character_histogram[i]
+			if new_character_histogram[i] < 0:
+				new_num_wildcards += new_character_histogram[i]
+				new_character_histogram[i] = 0
+				if new_num_wildcards < 0:
+					return None
+		return type(self)(character_histogram=new_character_histogram, num_wildcards=new_num_wildcards)
+
+	def find(self):
+		words_by_character_frequency = get_current_dictionary().words_by_character_frequency
+		anagrams = []
+		expanded_histograms = self.expand_wildcard()
+		for histogram in expanded_histograms:
+			if histogram in words_by_character_frequency:
```

Comment:
Rather than iterate over the whole dictionary to find one or more words (assuming that's what this is doing), binary search for one of them then look in front and behind for any other matches